Order for `Data.Bin`

mechvel opened the pull request:
You branch is updated according to «origin/master».

changed:      CHANGELOG.md
changed:      src/Data/Bin.agda
new file:     src/Data/Bin/Ordering.agda
changed:      src/Data/Bin/Properties.agda
changed:      src/Data/Nat/Properties.agda

mechvel commented:
This is the Bin-III pull request. 
1) Data.Bin.Ordering  is added.
2) A couple of proofs added to Data.Nat.Properties.
3) A couple of proof names changed in Data.Bin.Properties.

MatthewDaggitt commented:
Hi @mechvel, thanks for the next pull request.
- First of all, please can you undo **every** change in `Data.Bin.Properties` that is not renaming a proof. You've made an awful lot of formatting changes that go against the usual formatting in the library.
- As described in the style guide when we want to represent function application in a name then we use square brackets so `double x = 2 * x` would be named `double[x]=2x`. Please can you change every similar occurrence of this.
- Please can you split the induction proofs out into their own module `Data.Bin.Induction` like `Data.Nat.Induction` and `Data.Fin.Induction`.
- As discussed many times before, please can you make sure records are formatted properly. They're much better this time, but there are still a couple with ending brackets on the wrong line.
- As discussed many times before, please don't rename the reasoning primitives.

I'll then start looking at it in more detail.

mechvel commented:
Please, consider my commit of July 17.

>  First of all, please can you undo every change in Data.Bin.Properties that is not renaming a proof. 
> You've made an awful lot of formatting changes that go against the usual formatting in the library.

Done.
Only I have changed some comment headers. They are few. 
Please, consider them as a proposal, and change them as you like.

> As described in the style guide when we want to represent function application in a name then we 
> use square brackets so double ``x = 2 * x`` would be named ``double[x]=2x``. Please can you change > every  similar occurrence of this.

Fixed  (for  ``double-x, suc-x, acc-x, acc-..``).

>  Please can you split the induction proofs out into their own module Data.Bin.Induction like 
> Data.Nat.Induction and Data.Fin.Induction.

Done.

>  As discussed many times before, please can you make sure records are formatted properly. They're > much better this time, but there are still a couple with ending brackets on the wrong line.

Fixed.

> As discussed many times before, please don't rename the reasoning primitives.

Fixed.
As a result, there have appeared in Bin.Ordering many lines 
```
  where
  open IneqReasoning 
```
and the same for ``open ≤-Reasoning, open P.≡-Reasoning``.

mechvel commented:
Is not this better than ``2[1+_] and 1+[2_]`` ?
```
data Bin : Set where
  zero  :  Bin
  2∘1+ :  Bin → Bin    
  1+2*  :  Bin → Bin    
```

gallais commented on CHANGELOG.md:
```diff
@@ -41,7 +41,8 @@ The following new modules have been added to the library:
 
   Data.Bin
-  Data.Bin.Base
+  Data.Bin.Bas
```
Typo

mechvel replied:
Fixed

gallais commented on src/Data/Bin/Induction.agda:
```diff
@@ -0,0 +1,76 @@
+-- ("each number in ℕ is accessible from 0 by _<_").
+
+<-acc-ℕ {n} = Induction.Nat.<-wellFounded n
+
```
This should at most be a `private` alias. We should not be introducing
multiple names for the same thing.

mechvel replied:
This place is totally improved. See further notes.

MatthewDaggitt commented on CHANGELOG.md:
````diff
@@ -74,8 +76,19 @@ attached to all deprecated names to discourage their use.
+* In `Data.Bin.Properties`:
+  ```agda
+  2[1+-≢0  ↦  2[1+x]≢0
+  1+[2-≢0  ↦  1+2[1+x]≢0
````
These don't need to be listed as renamings as the originals have never been released.

mechvel replied:
Removed.

MatthewDaggitt commented on CHANGELOG.md:
````diff
@@ -74,8 +76,19 @@ attached to all deprecated names to discourage their use.
+
   ```
 
+  In several property names ``2, 2*`` replaced with ``double``.
````
Again doesn't need to be listed in the CHANGELOG

mechvel replied:
Removed

MatthewDaggitt commented on CHANGELOG.md:
```diff
@@ -84,11 +97,14 @@ Other minor additions
-  0≢1+n    : ∀ {n} → 0 ≢ suc n
-  n<1+n    : ∀ {n} → n < suc n
+  even≢odd     : ∀ m n → 2 * m ≢ suc (2 * n)
+  0≢1+n        : ∀ {n} → 0 ≢ suc n
```
We tend to leave implicit arguments out of types in the changelog.

mechvel replied:
these implicits are removed

MatthewDaggitt commented on src/Data/Bin/Induction.agda:
```diff
@@ -0,0 +1,76 @@
+open import Induction.WellFounded using (Acc; module Inverse-image; module Subrelation)
+open Acc using (acc)
+
+Acc< =  Acc _<_
```
I don't think any of these aliases help particularly. I think better to use the original names.

mechvel replied:
The whole fragment with aliases is replaced with 
```
<-acc {x} = Subrelation.accessible {_} {_} {_} {Bin} {_<_} {_<ₙon_} <⇒<ₙon <ₙon-acc-x
  where
  -- Derive accessibility for _<_ on Bin from accessibility for _<_ on ℕ:
  <ₙon-acc-x = Inverse-image.accessible {_} {_} {_} {Bin} {ℕ} {ℕ._<_} toℕ
                                        (Induction.Nat.<-wellFounded (toℕ x))
```

MatthewDaggitt commented on src/Data/Bin/Induction.agda:
```diff
@@ -0,0 +1,76 @@
+<-acc {x} =  Subrelation.accessible {_} {_} {_} {Bin} {_<_} {_<ₙon_} <⇒<ₙon (<ₙon-acc {x})
+
+acc[0] :  Acc< zero
+acc[0] =  acc (\x x<0 → contradiction x<0 ≮0)
```
Please use proper lambdas.

mechvel replied:
Done

MatthewDaggitt commented on src/Data/Bin/Induction.agda:
```diff
@@ -0,0 +1,76 @@
+-- The function  downFrom  below can also serve as an example of usage of Acc.
+------------------------------------------------------------------------------
+
+downFrom' : (x : Bin) → Acc _<_ x → List Bin
```
Please use `\prime` rather than `'`

mechvel replied:
Done.
What about local variable names in other places?
(it is easier to press a quote than to enter \prime).

MatthewDaggitt replied:
In local variables (i.e. names that aren't exposed) I guess it's fine. I'd still encourage the use of prime though for consistency. I know it's more difficult to type, it annoys me as well!

mechvel replied:
* I replaced a quote with \prime everywhere except in local variable names.
* It is not clear from your sentence whether a quote is allowed in local variables.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,791 @@
+open import Data.Sum using (_⊎_; inj₁; inj₂)
+open import Function using (_∘_; _$_; flip; case_of_; _on_)
+open import Level using () renaming (zero to 0ℓ)
+open import Relation.Binary using
```
I think with this number of imports we can drop the `using`

mechvel replied:
Done

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,791 @@
+open import Data.Bin.Base
+open import Data.Bin.Properties
+open import Data.Nat as ℕ using (ℕ; z≤n; s≤s)
+open import Data.Nat.Properties as ℕp using (m+n∸m≡n; m+n∸n≡m)
```
Please use subscript `p` here.

mechvel replied:
Done

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,791 @@
+
+infix 4 _<_ _>_ _≤_
+
+data _<_ : Rel Bin 0ℓ  where
```
Out of interest why have you defined the strict order first instead of the standard non-strict order? In every other data type we define `_≤_` first.

mechvel replied:
Suppose that it is needed to define (implement) for some domain D the relations
``≈, <, ≤``, where these relations make D a StrictTotalOrder for ``(≈, <)``,
and TotalOrder for ``(≈, ≤)``.
One can
(A) implement ``≈, <`` as basic, and define ``x ≤ y =  x < y ⊎ x ≈ y``
or
(B) implement ``≈, ≤`` as basic, and define ``x < y =  x ≤ y × ¬ x ≈ y``.
or
(C) implement them in some way specific to D
    (like, for example, Standard does with ℕ).

When you define something for a particular case, it is desirable to follow
some generic construction, if possible.
To my mind, the approach (A) is the most natural. Because

a) it is generic,

b) ``≤`` is often called "less or equal", so that it intends to express ``≤``
   as a disjunction of the two basics.

c) If one defines ``<`` via ``≤``, one still will have to prove
   ``x ≤ y ←→ x < y ⊎ x ≈ y`` (may be under some assuptions).  

d) the formula in (A) is simpler than the formula in (B).


Standard defines ``<`` for ℕ via ``≤`` and ``suc``.
But this is specific for ℕ: there are many TotalOrder domains having ``≈, ≤``
and without a sensible ``suc``.

As to me, I would rather define for ℕ
```
data _<_ ...
  where
  0<s : ∀ {n} → zero < suc n
  s<s : ∀ {m n} → m < n → suc m < suc n
...
m ≤ n =  m < n ⊎ m ≡ n 
```

In the sequel you write about x, y being not inferred from  x<y.
But in fact, they are inferred (see my replies below).

MatthewDaggitt replied:
Okay well if it doesn't result in any unsolved metavariables, I'm a little bit happier with it. Before discussing it further, please could we check this by removing any remaining explicitly given implicit arguments in the code. I've highlighted some of them in my latest review.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,791 @@
+For  odd<even (inj₁ x<y) :    1+2x < 2(1+y)  ~  2(1+x) ≤ 2(1+y)  ~  x ≤ y
+-}
+
+_>_ _≤_ _≥_ _≮_ _≯_ _≰_ _≱_ :  Rel Bin 0ℓ
```
Please could you add each of these above their definition. I know it's a bit longer but its much clearer to read.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,791 @@
+<-irrefl {2[1+ x ]} {2[1+ .x ]} refl =  x'≮x'
+  where
+  x'≮x' : 2[1+ x ] ≮ 2[1+ x ]
+  x'≮x' (even<even x<x) =  <-irrefl refl x<x
```
I think these are short enough that they can just be added to the same line, e.g. 
```
<-irrefl {2[1+ x ]} {2[1+ .x ]} refl (even<even x<x) =  <-irrefl refl x<x
```

mechvel replied:
Done (for the next pattern also).

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,791 @@
+                                                       odd<odd (<-trans x<y y<z)
+
+<⇒≯ : _<_ ⇒ _≯_
+<⇒≯ {x} {y} x<y y<x =  <-irrefl refl (<-trans {x} {y} {x} x<y y<x)
```
The fact that `x` and `y` aren't inferable here suggests that the definition of `<` isn't quite right.

mechvel replied:
In fact, they are inferrable
(if this means that it is possible to omit implicit arguments here). 
I have changed the implementations to 
```
<⇒≯ x<y y<x =  <-irrefl refl (<-trans x<y y<x)
...
>⇒≮ y<x =  <⇒≯ y<x
```
and they are type-checked all right.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,791 @@
+<-cmp zero     1+[2 _ ]  = tri< 0<odd  (λ()) ≮0
+<-cmp 2[1+ _ ] zero      = tri> (λ())  (λ()) 0<even
+<-cmp 2[1+ x ] 2[1+ y ]  with <-cmp x y
+... | tri< x<y _   _ =  tri< lt (<⇒≢ lt) (<⇒≯ lt)  where lt = even<even x<y
```
Why use a `where` here?

mechvel replied:
In order to avoid the three copies of the string "(even<even x<y)" in the right hand side.
Really, do you like to read it as repeated three times? (four times?)
Also the type checker (or compiler) may need, to discover that the term
```
  tri< (even<even x<y) (<⇒≢ (even<even x<y)) (<⇒≯ (even<even x<y))
```
is an instance of ``tri< Z (<⇒≢ Z) (<⇒≯ Z)``.
And it is easier to do this for the term
``tri< lt (<⇒≢ lt) (<⇒≯ lt)  where lt = even<even x<y``.

MatthewDaggitt replied:
That's a good reason, hadn't spotted that, apologies!

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,791 @@
+
+fromℕ-mono-< :  fromℕ Preserves ℕ._<_ ⟶ _<_
+fromℕ-mono-< {m} {n} m<n =
+  let
```
Please could you use `where` instead of `let`

mechvel replied:
```
    x       = fromℕ m;      y       = fromℕ n
    toℕ-x≡m = toℕ-fromℕ m;  toℕ-y≡n = toℕ-fromℕ n
  in
  case <-cmp x y of \
```
> Is there a reason why you're using case instead of with?

I have rewritten it to
```
fromℕ-mono-< {m} {n} m<n =  aux (<-cmp x y)
  where
  open P.≡-Reasoning
  x       = fromℕ m;      y       = fromℕ n
  toℕ-x≡m = toℕ-fromℕ m;  toℕ-y≡n = toℕ-fromℕ n

  aux : Tri (x < y) (x ≡ y) (x > y) → x < y
  aux (tri< x<y _   _) =  x<y
  ...
```
But I do not know how write this via ``with`` so that the four assignments
after ``where`` to be common for all branches of ``with``.
Can you demonstrate?

MatthewDaggitt replied:
Well I would usually use another `with`.
```
fromℕ-mono-< {m} {n} m<n with toℕ-fromℕ m | toℕ-fromℕ n
... | toℕ[x]≡m | toℕ[y]≡n = ...
```
but I recognise that may not be very efficient...

mechvel replied:
As you are in a doubt, I do not change this place.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,791 @@
+    x       = fromℕ m;      y       = fromℕ n
+    toℕ-x≡m = toℕ-fromℕ m;  toℕ-y≡n = toℕ-fromℕ n
+  in
+  case <-cmp x y of \
```
Is there a reason why you're using `case` instead of `with`?

mechvel replied:
See the previous item.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,791 @@
+  }
+
+<-resp-≈ :  _<_ Respects₂ _≡_
+<-resp-≈ =  ( (\{x y y'} y≡y' x<y → subst (x <_) y≡y' x<y) ,
```
Can you prove the left and right versions separately? (If we keep the definition of `_<_`)

mechvel replied:
Done

MatthewDaggitt commented on src/Data/Nat/Properties.agda:
```diff
@@ -561,6 +561,19 @@ m≤m+n (suc m) n = s≤s (m≤m+n m n)
+... | inj₁ m<n = inj₁ m''≤n'
+  where
+  m''≤n' : suc (suc m) ≤ suc n
+  m''≤n' = s≤s m<n
```
Again this proof can be removed from the `where`. You could even use `Sum.map` if you were feeling daring to get rid of the `with`.

mechvel replied:
Rewritten to ``... | inj₁ m<n = inj₁ (s≤s m<n)``.

MatthewDaggitt commented on src/Data/Bin/Properties.agda:
```diff
@@ -26,16 +26,16 @@ import Relation.Nullary.Decidable as Dec
-import Algebra.Properties.CommutativeSemigroup ℕₚ.*-semigroup ℕₚ.*-comm
-  as ℕ-*-semigroup
+import Algebra.Properties.CommutativeSemigroup ℕₚ.+-semigroup ℕₚ.+-comm as
+  Of+ℕ-semigroup
```
As requested, please get rid of these changes.

mechvel replied:
Returned
```
import Algebra.Properties.CommutativeSemigroup ℕₚ.*-semigroup ℕₚ.*-comm
  as ℕ-*-semigroup
```

MatthewDaggitt replied:
This change doesn't seem to have happened. Could you double check?

mechvel replied:
There was some confusion. I hope, now it agrees with the requirement.
Does it?

MatthewDaggitt replied:
No, unfortunately not.

MatthewDaggitt commented on src/Data/Bin/Properties.agda:
```diff
@@ -121,14 +121,14 @@ suc≢0 {2[1+ _ ]} ()
-suc-pred {zero}     0≢0  = contradiction refl 0≢0
-suc-pred {2[1+ _ ]} _    = refl
-suc-pred {1+[2 x ]} _    = sym (1+[2-as∘ x)
+suc-pred : ∀ x → x ≢ zero → suc (pred x) ≡ x
```
Why have you made this argument explicit?

mechvel replied:
By missing the point. Changed to implicit.

MatthewDaggitt commented on src/Data/Bin/Properties.agda:
```diff
@@ -150,7 +150,7 @@ toℕ-pred 2[1+ x ] =  cong ℕ.pred $ sym $ ℕₚ.*-distribˡ-+ 2 1 (toℕ x)
 
 toℕ-fromℕ : toℕ ∘ fromℕ ≗ id
-toℕ-fromℕ 0         = refl
+toℕ-fromℕ 0      = refl
```
Why have you misaligned this `=` sign?

mechvel replied:
I fix it now as usual, by taking the file with  ``emacs -nw`` and fixing the alignment. I hope it works. Does it?

MatthewDaggitt replied:
Yes. Thanks.

MatthewDaggitt commented on src/Data/Bin/Properties.agda:
```diff
@@ -215,16 +215,18 @@ toℕ-homo-+ 2[1+ x ] 2[1+ y ] = begin
                                           refl (toℕ x) (toℕ y) ⟩
   toℕ 2[1+ x ] ℕ.+ toℕ 2[1+ y ]      ∎
+  where
+  m = toℕ x;  n = toℕ y
```
Please could you remove these changes.

mechvel replied:
Done

MatthewDaggitt commented on src/Data/Bin/Properties.agda:
```diff
@@ -215,16 +215,18 @@ toℕ-homo-+ 2[1+ x ] 2[1+ y ] = begin
   toℕ (suc 2[1+ (x + y) ])              ≡⟨ toℕ-suc 2[1+ (x + y) ] ⟩
   ℕ.suc (toℕ 2[1+ (x + y) ])            ≡⟨⟩
-  ℕ.suc (2 ℕ.* (ℕ.suc (toℕ (x + y))))   ≡⟨ cong (λ v → ℕ.suc (2 ℕ.* ℕ.suc v)) (toℕ-homo-+ x y) ⟩
+  ℕ.suc (2 ℕ.* (ℕ.suc (toℕ (x + y))))   ≡⟨ cong (ℕ.suc ∘ (2 ℕ.*_) ∘ ℕ.suc) (toℕ-homo-+ x y) ⟩
```
Please could you remove this change, it's much less clear what the function is.

mechvel replied:
Done

MatthewDaggitt commented on src/Data/Bin/Properties.agda:
```diff
@@ -364,25 +367,22 @@ fromℕ-homo-+ (ℕ.suc m) n = begin
   }
 
-+-0-isMonoid : IsMonoid _+_ 0B
++-0-isMonoid : IsMonoid _+_ zero
```
Please could you put `0B` back?

mechvel replied:
Done

MatthewDaggitt replied:
Thanks, but I think you missed out `+-0-isCommutativeMonoid` and `+-0-monoid`.

mechvel replied:
Done.

MatthewDaggitt replied:
This change isn't appearing.

MatthewDaggitt commented on src/Data/Bin/Properties.agda:
```diff
@@ -602,7 +597,8 @@ fromℕ-homo-* m n = begin
 ------------------------------------------------------------------------
--- Structures
+-- Instances for Bin for some classical algebraic categories.
+-- Multiplicative part, CommutativeSemiring.
```
We've had this discussion twice before now. Please could you keep the standard section headers.

mechvel replied:
I copy your headers now.
By the way, which of these headers in your file are standard?
```
------------------------------------------------------------------------
-- Properties of _≡_
------------------------------------------------------------------------
```
,
```
------------------------------------------------------------------------
-- Properties of _+_
------------------------------------------------------------------------

------------------------------------------------------------------------
-- toℕ/fromℕ are homomorphisms for _+_
```
,
```
------------------------------------------------------------------------
-- Algebraic properties of _+_

-- Mostly proved by using the isomorphism between `ℕ` and `Bin` provided
-- by `toℕ`/`fromℕ`.
```
,

```
------------------------------------------------------------------------
-- Structures
```

MatthewDaggitt replied:
The standard ones are
```
------------------------------------------------------------------------
-- Properties of _≡_
------------------------------------------------------------------------

------------------------------------------------------------------------
-- Properties of _+_
------------------------------------------------------------------------

------------------------------------------------------------------------
-- Algebraic properties of _+_

------------------------------------------------------------------------
-- Structures
```

mechvel replied:
I hope, now it agrees with the style.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,791 @@
+  0<odd     : ∀ {x} → zero < 1+[2 x ]
+  even<even : ∀ {x y} → x < y → 2[1+ x ] < 2[1+ y ]
+  even<odd  : ∀ {x y} → x < y → 2[1+ x ] < 1+[2 y ]
+  odd<even  : ∀ {x y} → x < y ⊎ x ≡ y → 1+[2 x ] < 2[1+ y ]
```
This `⊎` is the reason why `x` and `y` aren't inferable from `x < y`. We should try to reformulate the definition to avoid this. I'd prefer to redefine this in terms of  `_≤_`, however if for some reason that's not possible then we need to split this case into two separate constructors.

mechvel replied:
x and y are inferrable here.
About defining ``≤``, see my response above.

MatthewDaggitt commented:
> As a result, there have appeared in Bin.Ordering many lines
> 
>   where
>   open IneqReasoning 

If you write `where open IneqReasoning` then it immediately cuts the number in half :)

> Is not this better than 2[1+_] and 1+[2_] ?
> data Bin : Set where
>   zero  :  Bin
>   2∘1+ :  Bin → Bin    
>   1+2*  :  Bin → Bin    

What are the advantages?

mechvel commented:
Please, consider my commit of July 20.
Each your note has been applied or answered.

I start now with filling in my replies.

mechvel commented:
Is not this better than ``2[1+]`` and ``1+[2]`` ?
```
    data Bin : Set where
    zero : Bin
    2∘1+ : Bin → Bin
    1+2* : Bin → Bin
```
> What are the advantages?

The advantage is that the constructor names do not contain any kind of parentheses.
Parentheses/brackets (sometimes unbalanced) are often used in variable names and in function names. I feel that allowing them in data constructor names brings more difficulty in choosing well-readable denotations.

mechvel commented:
I need Injective and Surjective as they are in Function.Definitions.
Recently I have seen Function.Definitions in ``master``.
But I do not find this module in the result of my last ``git pull``, that happened when I started a fork by new for Bin-III from ``master``.
What is the effect, I wonder.

MatthewDaggitt commented:
> The advantage is that the constructor names do not contain any kind of parentheses.
Parentheses/brackets (sometimes unbalanced) are often used in variable names and in function names. I feel that allowing them in data constructor names brings more difficulty in choosing well-readable denotations.

I agree that parantheses should be avoided if possible, but in this case I still think they're more readable than the alternatives. There shouldn't be any examples of where we use unbalanced parentheses in names. Can you point out where we do so and I'll change them?

The main problem I have with your suggestion above is that `2∘1+` makes it look like `2` is a function...

> I need Injective and Surjective as they are in Function.Definitions.
Recently I have seen Function.Definitions in master.
But I do not find this module in the result of my last git pull, that happened when I started a fork by new for Bin-III from master.
What is the effect, I wonder.

This is because they were added after your latest fork. To get them you need to run:
```
# pull the changes I've made to your code from the online version of your repository
git pull master
# fetch the changes to the standard library
git fetch upstream 
# bring your the changes to the standard library into your code
git pull upstream master
```

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+
+-- The ordering relation
+
+infix 4 _<_ _>_ _≤_
```
You're missing some of the relations here.

mechvel replied:
> I agree that parantheses should be avoided if possible, but in this case I still think they're 
> more readable than the alternatives. There shouldn't be any examples of where we use 
> unbalanced parentheses in names. Can you point out where we do so and I'll change 
> them?

I see the functions ``2[1+-as∘, 1+[2-as∘``. They are applied like this:
```
  suc (double x)   ≈⟨ sym (1+[2-as∘ x) ⟩
  1+[2 x ]         ∎
```
Now I rename them to ``2[1+_]-as∘, 1+[2_]-as∘``.
It this all right? 


>> infix 4 _<_ _>_ _≤_

> You're missing some of the relations here.

I add there ``_≮_ _≯_ _≰_ _≱_``.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+open import Relation.Nullary using (¬_; yes; no)
+open import Relation.Nullary.Negation using (contradiction)
+
+open Bin
```
I don't think this is necessary?

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+_≥_ =  flip _≤_
+
+_≮_ :  Rel Bin 0ℓ
+_≮_ x =  ¬_ ∘ _<_ x
```
It would be clearer what's going on if you wrote these out in terms of `x` and `y` and using the infix notation rather than using composition.

mechvel replied:
Done.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+_≱_ :  Rel Bin 0ℓ
+_≱_ x =  ¬_ ∘ _≥_ x
+
+≮0 : ∀ {x} → x ≮ zero    -- to use instead of λ()
```
This should be called `x≮0`

mechvel replied:
Done

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+<-irrefl {1+[2 x ]} {1+[2 .x ]} refl (odd<odd x<x)   =  <-irrefl refl x<x
+
+<⇒≢ :  _<_ ⇒ _≢_
+<⇒≢ {x} {_} x<y x≡y =  <-irrefl {x} {x} refl x<x
```
You can simplify this proof a lot by pattern matching on `x≡y`. This goes for many of the other proofs below as well. It would be worth trying it everywhere you use a propositional equality.

mechvel replied:
Replaced with ``<⇒≢ x<x refl =  <-irrefl refl x<x``

Pattern matching on `` x≡y`` is applied in several places
(though the proofs become less explicit and sometimes more difficult to understand
for a reader).

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+<-trans {_}      {zero}   {_}        ()
+<-trans {zero}   {_}      {1+[2 _ ]} _  _  =  0<odd
+
+<-trans {2[1+ x ]} {2[1+ y ]} {2[1+ z ]} (even<even x<y) (even<even y<z) =
```
You don't need to explicitly pattern match on the implicit arguments for this case and all the cases below. You should then be able to fit them all on one line :+1:

This is true almost everywhere where you're pattern matching on a proof of `_<_`.

mechvel replied:
Removed implicits in such places.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+
+<-cmp 1+[2 _ ] zero =  tri> (>⇒≮ gt) (>⇒≢ gt) gt
+  where
+  gt = 0<odd
```
I've added some guidance to the style guide that in the case of very short terms like these then they should go on the same line as the `where` in order to save space, i.e. `where gt = 0<odd`

mechvel replied:
I use this style feature now.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+  xN      = toℕ x
+  yN      = toℕ y
+  xN<yN   = toℕ-mono-< x<y
+  x'N<y'N = begin
```
This reasoning proof doesn't need to be in the `where`. Same with many examples below.

mechvel replied:
Fixed.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+≤→≢→< (inj₁ x<y) _   =  x<y
+≤→≢→< (inj₂ x≡y) x≢y =  contradiction x≡y x≢y
+
+0≤ :  ∀ x → zero ≤ x
```
Please could you add `x`s to all these names,

mechvel replied:
Done.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+  ; trans         = ≤-trans
+  }
+
+<-respˡ-≈ :  _<_ Respectsˡ _≡_
```
You should use `≡` in the names, not `≈`

mechvel replied:
Done

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+  }
+
+<-respˡ-≈ :  _<_ Respectsˡ _≡_
+<-respˡ-≈ {x} {_} =  subst (_< x)
```
You can use `respˡ` from `Relation.Binary.PropositionalEquality.Core` for these.

mechvel replied:
``<-resp*`` removed. Instead  ``Relation.Binary.PropositionalEquality.resp₂``
is used everywhere as ``(resp₂ _<_)``.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+
++-mono-<-≤ :  _+_ Preserves₂ _<_ ⟶ _≤_ ⟶ _<_
++-mono-<-≤ {x} {x'} {y} {y'} x<x' y≤y' =  begin-strict
+    x + y                     ≈⟨ sym $ cong₂ _+_ (fromℕ-toℕ x) (fromℕ-toℕ y) ⟩
```
Please could you better align this reasoning proof

mechvel replied:
Done

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+  1B + y    ≈⟨ sym (suc≗1+ y) ⟩
+  suc y     ∎
+  where
+  open IneqReasoning
```
Please put the `open` statements on the same line as the `where`, here and below.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+... | tri> _   _   1>x =  contradiction (x<1⇒x≡0 1>x) x≢0
+
+_<ₙon_ :  Rel Bin 0ℓ
+_<ₙon_ =  ℕ._<_ on toℕ
```
This should be defined at the top of the file with the other orderings.

mechvel replied:
Done

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+    pred 1B       ≈⟨ refl ⟩
+    zero          ∎
+
+x≢0-nor-1⇒1<x :  ∀ {x} → x ≢ zero → x ≢ 1B → 1B < x
```
This proof is a bit ugly and isn't used anywhere else. Could we remove it from this PR and if it happens we need it later, we can revisit it.

mechvel replied:
Removed.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+<-isStrictTotalOrder : IsStrictTotalOrder _≡_ _<_
+<-isStrictTotalOrder = record
+  { isEquivalence =  isEquivalence
+  ; trans         =  λ {x y z} → <-trans {x} {y} {z}
```
Are these implicits needed?

mechvel replied:
Now I remove such implicits everywhere. In early days Agda required them.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+
+≤-total :  Relation.Binary.Total _≤_
+≤-total x y  with <-cmp x y
+... | tri< le _  _  =  inj₁ (<⇒≤ {x} {y} le)
```
Again are these implicits needed?

mechvel replied:
Removed, as in many other places.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+  }
+
+<-strictTotalOrder : StrictTotalOrder _ _ _
+<-strictTotalOrder = record
```
For packages like these, there's no need to provide anything except the `is` proof.

mechvel replied:
Fixed everywhere.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+  open IneqReasoning
+
+x<2[1+x] :  ∀ x → x < 2[1+ x ]
+x<2[1+x] x =  <-trans {x} {suc x} {2[1+ x ]} (x<suc[x] x) (suc[x]<2[1+x] x)
```
Are these implicits needed?

mechvel replied:
Removed.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+
+double-back-mono-≤ :  ∀ {x y} → double x ≤ double y → x ≤ y
+double-back-mono-≤ {x} {y} 2x≤2y  with <-cmp x y
+... | tri< x<y _   _   =  <⇒≤ {x} {y} x<y
```
Are these implicits needed?

mechvel replied:
Implicits removed from ``*-monoʳ-< 1B {x} {y} x<y`` and ``<⇒≤ {x} {y} x<y``.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+  m≤m' = toℕ-mono-≤ x≤x';    n≤n' = toℕ-mono-≤ y≤y'
+
+*-monoʳ-≤ :  ∀ x → (x *_) Preserves _≤_ ⟶ _≤_
+*-monoʳ-≤ x {y} {y'} y≤y' =  *-mono-≤ {x} {x} {y} {y'} (≤-refl {x}) y≤y'
```
Are these implicits needed?

mechvel replied:
Removed

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+  y+x<y'+x' =  +-mono-<-≤ {y} {y'} {x} {x'} y<y' x≤x'
+
++-monoˡ-< :  ∀ x → (_+ x) Preserves _<_ ⟶ _<_
++-monoˡ-< x {y} {z} y<z =  +-mono-<-≤ {y} {z} {x} {x} y<z (≤-refl {x})
```
Are these implicits needed?

mechvel replied:
Removed

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,790 @@
+  m≤m' = toℕ-mono-≤ x≤x';   n≤n' = toℕ-mono-≤ y≤y'
+
++-monoˡ-≤ :  ∀ x → (_+ x) Preserves _≤_ ⟶ _≤_
++-monoˡ-≤ x {y} {z} y≤z =  +-mono-≤ {y} {z} {x} {x} y≤z (≤-refl {x})
```
Are these implicits needed?

mechvel replied:
Removed

mechvel commented:
In Data.Bin.Base we have now a definition  ``0B = zero``.                                              
What needs to be the rule for choosing between ``zero`` and ``0B`` in the further source?                                  
Set ``zero`` in samples, and set ``0B`` in other places?

mechvel commented:
Please, consider my commit(s) of July 30.

* First I applied all you notes to the code.
* Then I have done ``git pull`` and observed a huge change in Bin.Properties done by using RawMonoid homomorphism.
* I agreed my Bin.Properties  with yours by finding the marked differences and replacing my parts with  your parts. 
*  The following functions are added to Data.Bin.Base: ``toBits, fromBits, show``.
This is for conversion ``Bin <--> List Bit``  and to  show Bin as a string of bit characters.

MatthewDaggitt commented:
> In Data.Bin.Base we have now a definition 0B = zero.What needs to be the rule for choosing between zero and 0B in the further source?

If it's in a record and hence not being pattern matched on, I think it's clearer to use `0B`. If it is being pattern matched on then I'd use `zero`

> Then I have done git pull and observed a huge change in Bin.Properties done by using RawMonoid homomorphism.

Sounds like it worked, well done :+1:

> The following functions are added to Data.Bin.Base: toBits, fromBits, show.
This is for conversion Bin <--> List Bit and to show Bin as a string of bit characters.

Please don't add new content to the PR.  This PR should be just about adding the ordering relations. Please could you remove these functions.

> I have forgotten of using Injective and Surjective of Function.Definitions.
Need I to insert their usage now and do one more commit?

If you think it would be helpful then feel free to add another commit.

mechvel commented:
Please, consider my third commit of July 30.
It
* returns Data.Bin.Base to the previous state, as you required,
* fixes the definition of Function.Definitions.Core2.Surjective,
* sets the usage of Injective, Surjective into Data.Bin.Properties.

MatthewDaggitt commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+≤-reflexive : _≡_ ⇒ _≤_
+≤-reflexive {x} {_} refl =  ≤-refl {x}
+
+≤-trans : Transitive _≤_
```
This proof and several others can be found in `Relation.Binary.Construct.StrictToNonStrict`.

mechvel replied:
I set now
```
import Relation.Binary.Construct.StrictToNonStrict _≡_ _<_ as ToNonstrict
...
≤-trans = ToNonstrict.trans isEquivalence (resp₂ _<_) <-trans
...
≤-antisym = ToNonstrict.antisym isEquivalence <-trans <-irrefl
```
Other possible applications of ToNonstrict are not done because I have an
impression that they will not simplify the code.

> Thanks for removing the implicits, I would still kind of prefer that ``_<_`` was defined in 
> terms  of ``_≤_`` but upon reflection, it's probably not a hill that I want to die on so you'll 
> be happy to hear that the current definition can stay

Thank you very much!

mechvel commented on src/Data/Bin.agda:
```diff
@@ -9,5 +9,7 @@
+open import Data.Bin.Induction public
 open import Data.Bin.Properties public
   using (_≟_)
+open import Data.Bin.Ordering public
```
Are you sure? Bin/Properties will become of enormous size.

mechvel commented:
Please, consider my commit of August 5.
Most your notes are applied. Others will be now commented as replies.

mechvel commented on src/Data/Bin.agda:
```diff
@@ -9,5 +9,7 @@
 module Data.Bin where
 
 open import Data.Bin.Base public
+open import Data.Bin.Induction public
```
So, I remove this line now.

mechvel commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+open import Data.Product using (_,_; proj₁; proj₂)
+open import Data.Sum
+open import Function
+open import Level using () renaming (zero to 0ℓ)
```
Changed to ``open import Level using (0ℓ)``

mechvel commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+open import Level using () renaming (zero to 0ℓ)
+open import Relation.Binary
+open import Relation.Binary.PropositionalEquality as P
+open import Relation.Binary.PropositionalEquality.Core using (≢-sym; resp₂)
```
Removed.

mechvel commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+_≱_ :  Rel Bin 0ℓ
+x ≱ y =  ¬ (x ≥ y)
+
+_<ₙon_ :  Rel Bin 0ℓ
```
Renamed to ``_<ℕ_``

mechvel commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+<-irrefl {zero}     {zero}      _  ()
+<-irrefl {zero}     {2[1+ _ ]}  ()
+<-irrefl {zero}     {1+[2 _ ]}  ()
+<-irrefl {2[1+ _ ]} {zero}      ()
```
Removed.

mechvel commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+<-irrefl {zero}     {1+[2 _ ]}  ()
+<-irrefl {2[1+ _ ]} {zero}      ()
+<-irrefl {2[1+ x ]} {2[1+ .x ]} refl (even<even x<x) =  <-irrefl refl x<x
+<-irrefl {1+[2 x ]} {1+[2 .x ]} refl (odd<odd x<x)   =  <-irrefl refl x<x
```
Removed these implicits.

mechvel commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+<⇒≢ x<x refl =  <-irrefl refl x<x
+
+>⇒≢ :  _>_ ⇒ _≢_
+>⇒≢ {_} {_} y<x =  ≢-sym (<⇒≢ y<x)
```
Replaced with  ``>⇒≢ y<x =  ≢-sym (<⇒≢ y<x)``

mechvel commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+<-trans : Transitive _<_
+<-trans {_}             {_}      {zero}     _  ()
+<-trans {zero}          {_}      {2[1+ _ ]} _  _  =  0<even
+<-trans {_}             {zero}   {_}        ()
```
Removed.

mechvel commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+------------------------------------------------------------------------------
+
+toℕ-mono-< :  toℕ Preserves _<_ ⟶ ℕ._<_
+toℕ-mono-< {_}        {zero}     ()
```
Removed.

mechvel commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+  xN = toℕ x;  yN = toℕ y;  xN<yN = toℕ-mono-< x<y
+
+toℕ-mono-< {1+[2 x ]} {2[1+ y ]} (odd<even (inj₁ x<y)) =   begin
+  ℕ.suc (ℕ.suc (2 ℕ.* xN))    ≡⟨ refl ⟩
```
Replaced.

mechvel commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+------------------------------------------------------------------------------
+
+private
+  module IneqReasoning = InequalityReasoning {A = Bin} {_≈_ = _≡_} ≤-isPreorder <-trans
```
Renamed to  ``module ≤-Reasoning = ...``
(now almost everything is in Data.Bin.Properties).
But I failed to follow literally the declaration ``module ≤-Reasoning ...``
of Data.Integer.Properties. This is not type-checked.

Probably, you require to "hide the ``_≈_`` related syntax" in order to have in Data.Bin.Properties ``≡⟨ ... ⟩`` rather than ``≈⟨ ... ⟩``.
But I set ``≡⟨ ... ⟩`` everywhere, and it is type-checked.
?

mechvel commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+  double x          ∎
+  where open IneqReasoning;  px = pred x
+
+suc[x]≤⇒x< :  ∀ {x y} → suc x ≤ y → x < y
```
Done.

mechvel commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+    x             ≈⟨ sym (pred-suc x) ⟩
+    pred (suc x)  ≤⟨ pred-mono-≤ suc[x]≤1 ⟩
+    pred 1B       ≈⟨ refl ⟩
+    zero          ∎
```
Replaced with
```
x<1⇒x≡0 {x} x<1 =  x≤0⇒x≡0 $ begin
  x  ...     ≈⟨ sym (pred-suc x) ⟩
  ...
  zero       ∎
  where ...
```

There remains the place of line 857 
```
  aux (tri≈ _   x≡y _) =  contradiction m≡n (ℕₚ.<⇒≢ m<n)
    where
    open ≡-Reasoning
    m≡n = begin
      m       ≡⟨ sym toℕ-x≡m ⟩
      toℕ x   ≡⟨ cong toℕ x≡y ⟩
      toℕ y   ≡⟨ toℕ-y≡n ⟩
      n       ∎
```
Substituting the reasoning block for ``m≡n`` will make the code less clear.
Do this substituting?

Generally, what is the reason for prohibiting of a reasoning block after "where" ?

mechvel commented on src/Data/Bin/Ordering.agda:
```diff
@@ -0,0 +1,666 @@
+  ; antisym    = ≤-antisym
+  }
+
+≤-total :  Relation.Binary.Total _≤_
```
Replaced with ``≤-total : Total _≤_`` and inserted after ``≤-antisym``.

mechvel commented on src/Data/Bin/Properties.agda:
```diff
@@ -20,6 +20,8 @@ open import Data.Nat.Solver
 open import Data.Sum using (_⊎_; inj₁; inj₂)
 open import Function using (_∘_; _$_; id; const)
+open import Function.Definitions using (Injective)
+open import Function.Definitions.Core2 using (Surjective)
```
I fail with setting
```
open import Function.Definitions using (Injective; Surjective)
```
Because there are declared
```
toℕ-injective :  Injective {A = Bin} {B = ℕ} _≡_ _≡_ toℕ
...
toℕ-surjective :  Surjective {B = ℕ} _≡_ toℕ
```
, and the modules Function.Definitions and Function.Definitions.Core2 
have different parameters. 
All this leads to the report of
"ℕ !=< Bin of type Set
 when checking that the expression toℕ has type Rel ℕ _ℓ₂_415
"
for the declaration ``toℕ-surjective :  Surjective {B = ℕ} _≡_ toℕ``
And I do not guess what arguments needs the line ``toℕ-surjective : ...`` 
Please, advise.

mechvel commented on src/Data/Bin/Properties.agda:
```diff
@@ -29,45 +31,44 @@ open import Relation.Nullary.Negation using (contradiction)
 open import Algebra.Structures {A = Bin} _≡_
 import Algebra.Properties.CommutativeSemigroup ℕₚ.+-semigroup ℕₚ.+-comm
-  as ℕ-+-semigroup
+  as Of+ℕ-semigroup
```
This is also for your next message.
Please write explicitly what name do you suggest for this.
Meanwhile I set  
```
import Algebra.Properties.CommutativeSemigroup ℕₚ.+-semigroup ℕₚ.+-comm as Of+ℕ
```
, and the "*" variant occurs not used.

MatthewDaggitt commented:
Hi @mechvel, this afternoon I've gone through this and tidied everything up.

My one remaining question is in how you actually use the accessibility lemmas for `suc`, `pred`, `double` etc. in `Data.Bin.Induction`? Doesn't wrapping them up in the proofs hide the fact that the terms are structurally decreasing each time? I'd imagine therefore that they are not very useful when writing non-trivial recursive functions... Maybe I'm being slow, and if I am could you give an example of a proof that uses them?

mechvel commented:
> My one remaining question is in how you actually use the accessibility lemmas for suc, pred, 
> double etc. in Data.Bin.Induction? Doesn't wrapping them up in the proofs hide the fact that the 
> terms are structurally decreasing each time? I'd imagine therefore that they are not very useful 
> when writing non-trivial recursive functions... Maybe I'm being slow, and if I am could you give 
> an example of a proof that uses them?

1. The simplest example is   Data.Bin.Induction.downFrom.
If you demonstrate a simpler implementation for this function (4 patterns)
 without using ``<-acc``,  then let you replace it with the improved code.
(``Nat.suc`` makes a number structurally greater, but this is not so for ``Bin.suc``).

2. By the way, have you removed this ``downFrom`` ?
    This is a provision for making tests, for example, to run divMod, gcd on arguments taken from a 
    long list in order to test the performance.  I am not sure, but may be it is useful in standard library. 
    (?)

3. ``<-acc`` is used in Bin.DivMod.divMod,  Bin.GCD.gcd, which I am going to commit with further 
    PRs.  For example,  for ``a ≠ 0,  divMod a b``  is reduced to  ``divMod a' b,  where a' < a``,    
    and ``a'``  is not necessarily structurally less than ``a``.
When we come to Bin.DivMod, you would decide of whether <-acc is essential.

4. Other points.

* You have renamed ``2[1+_]-as..``  to ``2[1+_]-double-suc``.   All right.

* ``import ... as Of+ℕ``  changed to ``import ... as  ℕ-commutativeSemigroup`` 
  May be, it is better  ``Of-ℕ+commutativeSemigroup``.
Because 1) ``+`` shows that it is chosen the additive semigroup, not multiplicative,
2)  ``ℕ-commutativeSemigroup`` is understood and an instance of  CommutativeSemigroup
for ``ℕ``.  And this line means that it is imported a certain module (theory) parameterized by the instance of  CommutativeSemigroup and exporting various related items.

MatthewDaggitt commented:
> The simplest example is Data.Bin.Induction.downFrom (4 patterns of code).
If you demonstrate a simpler implementation for this function without using <-acc, then let you replace it with the improved code.

Sorry, I didn't mean `<-wellFounded` (renamed from `<-acc`). I meant the lemmas like `acc[suc[x]]⇒acc[x]` or `acc[double[x]]⇒acc[x]`. Do you have an example of using those? As I said above, I'm struggling to think of how you would use them.

> By the way, have you removed this downFrom ?
This is a provision for making tests, for example, to run divMod, gcd on arguments taken from a
long list in order to test the performance. I am not sure, but may be it is useful in standard library.
(?)

I did remove it. Sorry should have let you know. It wasn't relevant to this PR which is simply adding properties for `_<_`, and it was in the wrong place, as it's more a function over `List` than `Bin` (e.g. the existing `downFrom` is in `Data.List` not in `Data.Nat`).  If it proves useful later then we can add it back in a more appropriate place in a future PR.

> <-acc is used in Bin.DivMod.divMod, Bin.GCD.gcd, which I am going to commit with further
PRs. For example, for a ≠ 0, divMod a b is reduced to divMod a' b, where a' < a,
and a' is not necessarily structurally less than a.
When we come to Bin.DivMod, you would decide of whether <-acc is essential.

I haven't removed `<-acc`, I've simply renamed it `<-wellFounded` and made the parameter explicit. This brings it into line with the names of in other induction modules in the library.

> import ... as Of+ℕ changed to import ... as ℕ-commutativeSemigroup
May be, it is better Of-ℕ+commutativeSemigroup.

We've discussed this many times, both in this PR and the previous PR. Indeed we settled on `ℕ-+-semigroup` in the last PR, but you changed it back again in this PR. (which I forgot about in the last commit, and will change it back to now). The problem with `Of-ℕ+commutativeSemigroup` is that i) `Of` doesn't really add anything and ii) it doesn't follow the standard library convention of splitting up terms using `-`.

mechvel commented:
> Sorry, I didn't mean <-wellFounded (renamed from <-acc). I meant the lemmas like 
> acc[suc[x]]⇒acc[x] or acc[double[x]]⇒acc[x]. Do you have an example of using those? As I said 
> above, I'm struggling to think of how you would use them.

It occurs that the current PR modules do not use anything from ``Bin.Induction`` except ``<-acc``.
1) Some of these lemmas are used in Bin.Logarithm.  But I do not know of whether you would agree to include Bin.Logarithm in standard.
2) I think, these lemmas are useful in general. For example, the type
```
  (x : Bin) (x≢0 : x ≢ 0#) (acc-2x : Acc< (2* x)) →
  let 2x≢0  = x≢0⇒2x≢0 x≢0
       acc-x = acc-2x⇒acc-x acc-2x
  in
  log₂-rec (2* x) 2x≢0 acc-2x  ≡  1+ (log₂-rec x x≢0 acc-x)
```
expresses the statement   ``log (2B * x) ≡  1+ (log x)`` for a nonzero (x : Bin).
Here ``acc-x`` and ``acc-2x`` are needed to prove termination in applying  log to different arguments.

Let you decide yourself whether to remove these lemmas.  
 
* About removed ``Bin.Induction.downFrom``.
> It wasn't relevant to this PR which is simply adding properties for _<_, and it was in the wrong 
> place, as it's more a function over List than Bin (e.g. the existing downFrom is in Data.List not 
> in Data.Nat). 

Data.List.Base.downFrom  cannot serve for ``List Bin``. Because the latter needs to be based on ``Bin.pred``, which is fast. And Data.List.Base.downFrom is based on a numeration ``ℕ → A``,
which is slow for A = Bin, because it presumes initial applying  ``toℕ``. 
Generally, downFrom has sense for (List A), where A has \bottom, a well-founded ordering ``_<_ ``
and an operation predLike : A -> A with the property  ``predLike x < x`` for all x > \bottom.

Currently let you decide yourself whether to remove this downFrom.

MatthewDaggitt commented:
Okay, I've removed the accessibility lemmas and `downFrom`, if and when you add `log` then we can reinclude them if they turn out to be necessary. Also I think you misunderstood my point about `downFrom`. I was saying it's placement was wrong, not it's implementation.

> The problem with import ... as ℕ-+-commutativeSemigroup
is that the reader will see in the code open ℕ-+-commutativeSemigroup using (foo)
and will think that ℕ-+-commutativeSemigroup is an instance of CommutativeSemigroup.
And the intended meaning is: the item foo imported from a certain module parameterized by
Nat.Properties.+-commutativeSemigroup.

That's a reasonable point, I've renamed it `ℕ-+-semigroupProperties`.

Right, merging in!

mechvel commented:
> Also I think you misunderstood my point about downFrom. I was saying it's placement was 
> wrong, not it's implementation.

I wrote that downFrom for Bin cannot be placed in  ``Data/List/...`` because its definition is specific to the type Bin. Probably, either it needs to be somewhere in ``Bin/`` 
or it needs to be generalized, and then it would fit ``Data/List/...``.

mechvel commented:
I recall a certain item:
thank you for introducing to Bin.Properties structures and proofs induced by _morphisms_.
This way it is nicer (I did not know about existence of a morphism tool in standard library).

MatthewDaggitt commented:
> thank you for introducing to Bin.Properties structures and proofs induced by morphisms.
This way it is nicer (I did not know about existence of a morphism tool in standard library).

No worries. It was only recently added :+1:
